src/condor_shadow.V6.1/baseshadow.cpp

```diff
@@ -936,13 +937,38 @@ BaseShadow::evictJob( int exit_reason, const char* reason_str, int reason_code,
 
 	// Note: improveReasonAttributescan change the reason_code and subcode,
 	// and will pass back a potentially improved reason string.
+	std::string url_file_type; // This will be used to store the protocol used for file transfer.
 	std::string improved_reason_str =
-		improveReasonAttributes(reason_str, reason_code, reason_subcode);
+		improveReasonAttributes(reason_str, reason_code, reason_subcode, url_file_type);
 	// If we have an improved reason string, use it instead.
 	if (!improved_reason_str.empty()) {
 		reason_str = improved_reason_str.c_str();
 	}
 
+	// If we had a problem moving files, we want to record the protocol used
+	// for the file transfer, so that we can update the Evict Num Reasons
+	// attributes by protocol in the schedd.
+	if (reason_code == FILETRANSFER_HOLD_CODE::DownloadFileError ||
+		reason_code == FILETRANSFER_HOLD_CODE::UploadFileError)
+	{
+		if (url_file_type.empty()) {
+			// If we don't have a URL file type, then we don't know what protocol was used.
+			// So just set the protocol to "unknown".
+			url_file_type = "Cedar";
+		} else {
+			// Make url_file_type camel case, so that it is consistent with the
+			// other protocols we use in the Additional Hold/Evict Num Reasons attributes.
+			std::transform(url_file_type.begin(), url_file_type.end(), url_file_type.begin(),
```

Comment:
Not going to hold the PR back on this, but using ranges variants look much cleaner (i.e. `std::ranges::transform(...)`